modules/core/src/test/java/org/apache/ignite/internal/processors/cache/GridCacheRebalanceOrderTest.java

```diff
@@ -0,0 +1,177 @@
+    private static final long REBALANCE_TIMEOUT = 5_000;
+
+    /** Flag indicates that local listener should be used to track rebalance events. */
+    private boolean trackRebalanceEvts;
+
+    /** Caches rebalance events. */
+    private final List<CacheRebalancingEvent> evtsList = Collections.synchronizedList(new ArrayList<>());
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        if (trackRebalanceEvts) {
+            Map<IgnitePredicate<? extends Event>, int[]> listeners = new HashMap<>();
+
+            listeners.put(new IgnitePredicate<CacheRebalancingEvent>() {
+                @Override public boolean apply(CacheRebalancingEvent evt) {
+                    evtsList.add(evt);
+
+                    return true;
+                }
+            }, new int[] {EVT_CACHE_REBALANCE_STOPPED});
+
+            cfg.setLocalEventListeners(listeners);
+
+            cfg.setIncludeEventTypes(EventType.EVTS_ALL);
+        }
+
+        return cfg;
+    }
+
+    /**
+     * Stops all nodes after test.
+     */
+    @After
+    public void testCleanup() {
+        stopAllGrids();
+    }
+
+    /**
+     * Tests that caches with rebalance mode equals to SYNC are rebalanced in the first place.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testRebalanceOrderBasedOnCacheRebalanceMode() throws Exception {
+        Ignite g = startGrid(0);
+
+        // Fix the expected order of rebalance.
+        List<IgniteBiTuple<Integer, CacheRebalanceMode>> order = new ArrayList<>();
+        order.add(new T2<>(0, SYNC));
+        order.add(new T2<>(0, SYNC));
+        order.add(new T2<>(0, ASYNC));
+        order.add(new T2<>(0, ASYNC));
+        order.add(new T2<>(1, SYNC));
+        order.add(new T2<>(1, SYNC));
+        order.add(new T2<>(1, ASYNC));
+        order.add(new T2<>(1, ASYNC));
+
+        // Prepare caches with different rebalance mode and order.
+        List<IgniteCache<Integer, Integer>> caches = new ArrayList<>();
+        for (int i = order.size() - 1; i >= 0; i--) {
+            int rebalanceOrder = order.get(i).get1();
+
+            CacheRebalanceMode rebalanceMode = order.get(i).get2();
+
+            caches.add(g.getOrCreateCache(getCacheConfiguration(
+                "cache-" + i + "-order-" + rebalanceOrder + "-mode-" + rebalanceMode,
+                rebalanceOrder,
+                rebalanceMode)));
+        }
+
+        // Fill values.
+        for (IgniteCache<Integer, Integer> c : caches)
+            c.put(12, 21);
+
+        trackRebalanceEvts = true;
+
+        Ignite g1 = startGrid(1);
+
+        // Wait for all rebalance futures.
+        for (IgniteCache<Integer, Integer> c : caches)
+            grid(1).context().cache().internalCache(c.getName()).preloader().syncFuture().get(REBALANCE_TIMEOUT);
```

Comment:
`syncFuture` is for preloading only on a node startup, let's use `awaitPartitionMapExchange`

Reply:
Yep, and that is the case! :)
The most important impact of the proposed change is "unblocking" a user thread that is starting Ignite instance.
As you know, `Ignition.start()` blocks the thread until the node joined the cluster and all `SYNC` caches are rebalanced.
So, the fix allows us to "speed up" the start of the node, from the user point of view.

Reply:
As far I can see, the `evtsList` may not contains all of the rebalanced caches since at the moment of the current thread will be released and we start checking for cache order (in the loop above) the rebalance still may not be finished for ASYNC caches. Right?

Reply:
`evtsList` should contain all events after the following loop:

```
        // Wait for all rebalance futures.
        for (IgniteCache<Integer, Integer> c : caches)
            grid(1).context().cache().internalCache(c.getName()).preloader().syncFuture().get(REBALANCE_TIMEOUT);
```

In other words, `syncFuture().get()` guarantees that initial rebalancing for the cache is already completed. It does not matter which cache rebalancing mode is used - `SYNC` or `ASYNC`.

Reply:
Yes, you're right. Agree